File: leetcode/letter-combinations-of-a-phone-number.py
# -*- coding: utf-8 -*-
# @Author: Tairan Gao
# @Date:   2023-08-05 08:23:15
# @Last Modified by:   Tairan Gao
# @Last Modified time: 2023-08-05 08:37:56

from typing import List


class Solution:
    def __init__(self) -> None:
        self.letterDict = {
            2: ["a", "b", "c"],
            3: ["d", "e", "f"],
            4: ["g", "h", "i"],
            5: ["j", "k", "l"],
            6: ["m", "n", "o"],
            7: ["p", "q", "r", "s"],
            8: ["t", "u", "v"],
            9: ["w", "x", "y", "z"]
        }
        self.res = []
    
    def letterCombinations(self, digits: str) -> List[str]:
        self.backTrack(digits, 0, "")

        return self.res
    
    def backTrack(self, digits, ind, cur):
        if (ind == len(digits)):
            if cur:
                self.res.append(cur)
            return
        
        num = int(digits[ind])
        for letter in self.letterDict[num]:
            cur += letter
            self.backTrack(digits, ind + 1, cur)
            cur = cur[:-1]

    

if __name__ == "__main__":
    test = Solution()
    print(test.letterCombinations("23"))
    

File: leetcode/letter-combinations-of-a-phone-number.cpp
/*
 * @Author: Tairan Gao
 * @Date:   2023-08-05 08:38:30
 * @Last Modified by:   Tairan Gao
 * @Last Modified time: 2023-08-05 09:18:38
 */

#include <vector>
#include <string>
#include <iostream>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    vector<string> letterCombinations(string digits)
    {
        vector<string> res;
        if (digits.empty())
            return res;

        unordered_map<int, string> dict;
        dict = {{2, "abc"}, {3, "def"}, {4, "ghi"}, {5, "jkl"}, {6, "mno"}, {7, "pqrs"}, {8, "tuv"}, {9, "wxyz"}};
        string cur(digits.size(), ' ');
        backTrack(digits, 0, cur, dict, res);

        return res;
    }

    void backTrack(string &digits, int idx, string cur, unordered_map<int, string> &dict, vector<string> &res)
    {
        if (cur.back() != ' ')
        {
            res.push_back(cur);
            return;
        }

        int num = digits[idx] - '0';
        for (auto const &letter : dict[num])
        {
            cur[idx] = letter;
            backTrack(digits, idx + 1, cur, dict, res);
        }
        return;
    }
};

int main()
{
    Solution s;
    string digits = "23";
    vector<string> res = s.letterCombinations(digits);

    for (auto const &str : res)
    {
        cout << str << endl;
    }

    return 0;
}